examples/step-57/step-57.cc

```diff
@@ -0,0 +1,878 @@
+  template <int dim>
+  class BoundaryValues : public Function<dim>
+  {
+  public:
+    BoundaryValues() : Function<dim>(dim+1) {}
+    virtual double value(const Point<dim> &p,
+                         const unsigned int component) const;
+
+    virtual void   vector_value(const Point <dim>    &p,
+                                Vector<double> &values) const;
+  };
+
+  template <int dim>
+  double BoundaryValues<dim>::value(const Point<dim> &p,
+                                    const unsigned int component) const
+  {
+    Assert (component < this->n_components,
+            ExcIndexRange (component, 0, this->n_components));
+    if (component == 0 && std::abs(p[dim-1]-1.0) < 1e-10)
+      return 1.0;
+
+    return 0;
+  }
+
+  template <int dim>
+  void BoundaryValues<dim>::vector_value (const Point<dim> &p,
+                                          Vector<double>   &values) const
+  {
+    for (unsigned int c = 0; c < this->n_components; ++c)
+      values(c) = BoundaryValues<dim>::value (p, c);
+  }
+
+
+  // @sect3{BlockSchurPreconditioner for Navier Stokes equations}
+  //
+  // The block
+  // Schur complement preconditioner is defined in this part. As discussed in
+  // the introduction, the preconditioner in Krylov iterative methods is
+  // implemented as a matrix-vector product operator. In practice, the Schur
+  // complement preconditioner is decomposed as a product of three matrices(as
+  // presented in the first section).  The $\tilde{A}^{-1}$ in the first factor
+  // involves a solve for the linear system $\tilde{A}x=b$. Here we solve
+  // this system via a direct solver for simplicity. The computation involved
+  // in the second factor is a simple matrix-vector multiplication. The Schur
+  // complement $\tilde{S}$ can be well approximated by the pressure mass
+  // matrix and its inverse can be obtained through an inexact solver. Because
+  // the pressure mass matrix is symmetric and positive definite, we can use
+  // CG to solve the corresponding linear system.
+  //
+  template <class PreconditionerMp>
+  class BlockSchurPreconditioner : public Subscriptor
+  {
+  public:
+    BlockSchurPreconditioner (double                                     gamma,
+                              double                                     viscosity,
+                              const BlockSparseMatrix<double>            &S,
+                              const SparseMatrix<double>                 &P,
+                              const PreconditionerMp                     &Mppreconditioner
+                             );
+
+    void vmult (BlockVector<double>       &dst,
+                const BlockVector<double> &src) const;
+
+  private:
+    const double gamma;
+    const double viscosity;
+    const BlockSparseMatrix<double> &stokes_matrix;
+    const SparseMatrix<double> &pressure_mass_matrix;
+    const PreconditionerMp &mp_preconditioner;
+    SparseDirectUMFPACK A_direct;
```

Comment:
Might be nicer to call it `A_inverse`.

You so nicely line up member declarations in the main class. Maybe do the same here?